### ISSUE TITLE
[Feature Request] Add the ability to determine if the onscreen keyboard is currently visible. (Possible events?)

### ISSUE THREAD

**Issue by me4502** [libgdx/libgdx#3892]
Currently if a fullscreen table is used, it's hard to offset the contents when an onscreen keyboard shows up, as there isn't a way of knowing whether it is there or not.

If a way to determine if it's there or not was added, or even possible onKeyboardShow, and onKeyboardHide, it'd allow this to be done.

**Comment by Tom-Ski** [comment]
Oops.  I'm not entirely sure if this is actually possible on (on android at least).  Maybe in later versions of Android there is a direct way, perhaps @MobiDevelop knows more.

**Comment by MobiDevelop** [comment]
I don't believe it is possible while having the app/game full screen. When fullscreen, the keyboard overlays the surface, rather than resizing it to make room for the keyboard.

**Comment by me4502** [comment]
I'm not familiar with android, but doesn't it have a system like iOS where it scales the screen when the keyboard comes up?

**Comment by MobiDevelop** [comment]
Not in the case of the app requesting fullscreen, which all libgdx apps do by default. You can, of course, change that and then listen for the resizes yourself.

**Comment by pbertsch** [comment]
Hi,

I would sponsor the Software Keyboard check to AndroidOnscreenKeyboard.java
Would you agree ?
```
/**
	 * Check if the Software Keyboard is visible or not.
	 * @param v can be NULL
	 * @return true if visible, false otherwise
	 */
	public Boolean isKeyboadVisible(View v){

		InputMethodManager imm = (InputMethodManager) context
				.getSystemService(Context.INPUT_METHOD_SERVICE);

		if(v != null){
			return imm.isActive(v);
		}

		if (imm.isAcceptingText()) {
			Log.d("Keyboard","Software Keyboard was shown");
			return true;
		} else {
			Log.d("Keyboard", "Software Keyboard was not shown");
			return false;
		}
	}
```

and a new method into AndroidInput.java

```
/**
	 * Check if te Software Keyboard is visible or not
 	 * @param view
	 * @return
	 */
	public Boolean isOnscreenKeyboardVisible(final View view){
		return this.onscreenKeyboard.isKeyboadVisible(view);
	}

```

**Comment by intrigus** [comment]
@pbertsch This doesn't work, I tested it.
See http://stackoverflow.com/a/31090451 and http://stackoverflow.com/a/27567074 for a "better" solution (All solutions are quite hacky).

@me4502 Shouldn't you know when you show/hide the softkeyboard via `Gdx.input.setOnscreenKeyboardVisible(false)`?
But you are probably talking about e.g. TextField which itself shows the softkeyboard. 
So you probably want a listener which fires, when `Gdx.input.setOnscreenKeyboardVisible(value)` is called, right?

If not, could you please explain what you mean?

**Comment by me4502** [comment]
A listener for that would work, yes. I just want to be able to manipulate the screen so that stuff is moved up when the keyboard is visible.

**Comment by calanti** [comment]
I have been experimenting with this and found a pretty nice solution that works well on all my Android testing devices. The brilliant thing about it is it doesn't actually do anything with the keyboard, so it avoids all the issues you might find with different keyboards. What it does though is monitors the Android's visible view and provides you the dimensions, so you can work out the height of the keyboard and work around that.

[This](http://stackoverflow.com/questions/9604963/how-do-libgdx-detect-keyboard-presence/33188659#33188659) is what you want to implement.

Note I found I had to scale the returned view dimensions by my virtual/real screen scale for consistency on lower resolutions, if you have forced your game to a specific resolution, and create a small work around so I could launch the Desktop, other than that you can implement it as is.

What I then did is create an extended class of the TextField, with this;

`    @Override
    public void act(float delta) {
        maintainVisibility();
        super.act(delta);
    }

    private void maintainVisibility(){
        if(this.getStage() instanceof StageExt){
            StageExt stage = (StageExt) this.getStage();
            if(super.justFocused()){ 
                float y = this.localToStageCoordinates(new Vector2(0, 0)).y;
                if(y < stage.getKeyboardHeight()+50){
                    stage.interpolateCamTo(0, stage.getKeyboardHeight()-y+50, 0.5f, Interpolation.swingOut);
                } else {
                    stage.interpolateCamTo(0, 0, 0.5f, Interpolation.swingOut);
                }
            }

        }
    }`

Basically, what this does is if the TextField position is less than the keyboard height (and therefore hidden behind it) , the screen bounces up to make the TextField visible.

Note this is actually an extended class of another extended TextField class, and also I have my own extended Stage class, so it still won't work for you, but the methods I incorporated in these are fairly simple.

super.justFocused() is basically a 1 time boolean to capture the single frame when the TextField was clicked, 

stage.getKeyBoardHeight() is similar to the original link posted above.

stage.interpolateCamTo() just animates the camera position into place so it looks nice.

It is then my Stage class that handles the closing of the Keyboard event. You can't rely on the TextField loosing focus here because if the user closes the keyboard with the back button the TextField is still focused. This is impossible to solve or work around because Android does not pass the back button event to libGDX (yay)

Hope this helps you and others. It would be nice if libGDX could incorporate this Android visible view into the backend, so we don't need to modify all the application launchers to suit.
